sharding/contracts/sharding_manager.sol

```diff
@@ -1,253 +1,260 @@
-    require(_number == headerVars.score);
-
-    // Add the header
-    collationHeaders[_shardId][headerVars.entireHeaderHash] = CollationHeader({
-      parentHash: _parentHash,
-      score: headerVars.score
+    notaryRegistry[notaryAddress] = Notary({
+      deregisteredPeriod: 0,
+      poolIndex: index,
+      deposited: true
     });
 
-    // Update the latest period number
-    periodHead[_shardId] = int(_expectedPeriodNumber);
-
-    // Determine the head
-    if (headerVars.score > collationHeaders[_shardId][shardHead[_shardId]].score) {
-      shardHead[_shardId] = headerVars.entireHeaderHash;
-      headerVars.isNewHead = true;
+    // if current index is greater than notary sample size, increase notary sample size for next period
+    if (index >= nextPeriodNotarySampleSize) {
+      nextPeriodNotarySampleSize = index + 1;
     }
 
-    CollationAdded(_shardId, _expectedPeriodNumber, _periodStartPrevHash,
-                   _parentHash, _transactionRoot, _coinbase, _stateRoot, 
-                   _receiptRoot, _number, headerVars.isNewHead, headerVars.score);
+    NotaryRegistered(notaryAddress, index);
+    return true;
+  }
+
+  function deregisterNotary() public returns(bool) {
+    address notaryAddress = msg.sender;
+    uint index = notaryRegistry[notaryAddress].poolIndex;
+    require(notaryRegistry[notaryAddress].deposited);
+    require(notaryPool[index] == notaryAddress);
+
+    updateNotarySampleSize();
+    
+    uint deregisteredPeriod = block.number / PERIOD_LENGTH;
+    notaryRegistry[notaryAddress].deregisteredPeriod = deregisteredPeriod;
 
+    stackPush(index);
+    --notaryPoolLength;
+    NotaryDeregistered(notaryAddress, index, deregisteredPeriod);
     return true;
   }
 
-  // Records a request to deposit msg.value ETH to address to in shard shard_id
-  // during a future collation. Saves a `receipt ID` for this request,
-  // also saving `msg.sender`, `msg.value`, `to`, `shard_id`, `startgas`,
-  // `gasprice`, and `data`.
-  function txToShard(address _to, int _shardId, uint _txStartgas, uint _txGasprice,
-                     bytes12 _data) public payable returns(int) {
-    receipts[numReceipts] = Receipt({
-      shardId: _shardId,
-      txStartgas: _txStartgas,
-      txGasprice: _txGasprice,
-      value: msg.value,
-      sender: msg.sender,
-      to: _to,
-      data: _data
-    });
-    var receiptId = numReceipts;
-    ++numReceipts;
+  function releaseNotary() public returns(bool) {
+    address notaryAddress = msg.sender;
+    uint index = notaryRegistry[notaryAddress].poolIndex;
+    require(notaryRegistry[notaryAddress].deposited == true);
+    require(notaryRegistry[notaryAddress].deregisteredPeriod != 0);
+    require((block.number / PERIOD_LENGTH) > (notaryRegistry[notaryAddress].deregisteredPeriod + NOTARY_LOCKUP_LENGTH));
 
-    TxToShard(_to, _shardId, receiptId);
-    return receiptId;
+    delete notaryRegistry[notaryAddress];
+    notaryAddress.transfer(NOTARY_DEPOSIT);
```

Comment:
Shouldn't we use `.send` and check that the call was successful? 
Seems like good practice, but I am not sure